src/libstore/s3-binary-cache-store.cc

```diff
@@ -184,6 +186,75 @@ S3Helper::FileTransferResult S3Helper::getObject(
     return res;
 }
 
+/* An adapter that can be used to convert a Sink to ostream or iostream.
+
+   Example:
+
+   SinkToBidirectionalDeviceAdapter deviceToSink(sink);
+   boost::iostreams::stream<SinkToBidirectionalDeviceAdapter> streamToSink(deviceToSink);
+
+ */
+struct SinkToBidirectionalDeviceAdapter {
+    typedef char char_type;
+    typedef boost::iostreams::bidirectional_device_tag category;
+    Sink & sink;
+    std::streamsize totalSize;
+
+    SinkToBidirectionalDeviceAdapter(Sink & sink) : sink(sink), totalSize(0) {}
+
+    std::streamsize read(char* data, std::streamsize dataSize)
+    {
+        // Intentionally reports no data read. This struct is only intended for sinks.
+        return 0;
+    }
+
+    std::streamsize write(const char* data, std::streamsize dataSize)
+    {
+        auto strViewData = std::string_view(data, dataSize);
+        this->sink(strViewData);
+        this->totalSize += dataSize;
+        return dataSize;
+    }
+};
+
+S3Helper::FileStreamResult S3Helper::getObjectStreamed(
+    const std::string & bucketName, const std::string & key, Sink & sink)
+{
+    debug("streaming from 's3://%s/%s'...", bucketName, key);
+
+    auto request =
+        Aws::S3::Model::GetObjectRequest()
+        .WithBucket(bucketName)
+        .WithKey(key)
+        .WithResponseContentEncoding("identity");
```

Comment:
Since we are streaming, we would need to get the content encoding before the download starts. Unfortunately, we were not able to find a way to inject the content encoding into the `streamtoSink` object before the streaming starts.

This is why we explicitly set the encoding to "identity" and are not applying the `decompress` step (unlike the old `getObject` function in L174 above).

Note that this is separate from nix's own compression of nar archives. Nix will still decompress nar archives if they were compressed when it uploaded them to S3.

Reply:
My intuition is that it's a reasonable tradeoff since Nix already does its own compression most of the time, and most people will only use an `s3://` binary cache from inside aws (binary caches hosted on s3 but used outside of aws's datacenters will generally get an http proxy on front of them), so on a fast-enough network connection. @edolstra do you think there's a significant drawback in not having any compression on the transport layer here?